exchanges/gateio/gateio_websocket.go

```diff
@@ -12,720 +15,943 @@ import (
-	err := g.Websocket.Conn.Dial(&dialer, http.Header{})
+	err = g.Websocket.Conn.Dial(&dialer, http.Header{})
 	if err != nil {
 		return err
 	}
-
+	pingMessage, err := json.Marshal(WsInput{
+		ID:      g.Websocket.Conn.GenerateMessageID(false),
+		Time:    time.Now().Unix(),
+		Channel: spotPingChannel,
+	})
+	if err != nil {
+		return err
+	}
+	g.Websocket.Conn.SetupPingHandler(stream.PingHandler{
+		Websocket:   true,
+		Delay:       time.Second * 15,
+		Message:     pingMessage,
+		MessageType: websocket.PingMessage,
+	})
 	g.Websocket.Wg.Add(1)
-	go g.wsReadData()
+	go g.wsReadConnData()
+	return nil
+}
 
-	if g.IsWebsocketAuthenticationSupported() {
-		err = g.wsServerSignIn(context.TODO())
+func (g *Gateio) generateWsSignature(secret, event, channel string, dtime time.Time) (string, error) {
+	msg := "channel=" + channel + "&event=" + event + "&time=" + strconv.FormatInt(dtime.Unix(), 10)
+	mac := hmac.New(sha512.New, []byte(secret))
+	if _, err := mac.Write([]byte(msg)); err != nil {
+		return "", err
+	}
+	return hex.EncodeToString(mac.Sum(nil)), nil
+}
+
+// wsReadConnData receives and passes on websocket messages for processing
+func (g *Gateio) wsReadConnData() {
+	defer g.Websocket.Wg.Done()
+	for {
+		resp := g.Websocket.Conn.ReadMessage()
+		if resp.Raw == nil {
+			return
+		}
+		err := g.wsHandleData(resp.Raw)
 		if err != nil {
 			g.Websocket.DataHandler <- err
-			g.Websocket.SetCanUseAuthenticatedEndpoints(false)
-		} else {
-			var authsubs []stream.ChannelSubscription
-			authsubs, err = g.GenerateAuthenticatedSubscriptions()
-			if err != nil {
-				g.Websocket.DataHandler <- err
-				g.Websocket.SetCanUseAuthenticatedEndpoints(false)
-			} else {
-				err = g.Websocket.SubscribeToChannels(authsubs)
+		}
+	}
+}
+
+// wsReadData read coming messages thought the websocket connection and pass the data to wsHandleData for further process.
+func (g *Gateio) wsReadData() {
+	defer g.Websocket.Wg.Done()
+	for {
+		select {
+		case <-g.Websocket.ShutdownC:
+			select {
+			case resp := <-responseStream:
+				err := g.wsHandleData(resp.Raw)
 				if err != nil {
-					g.Websocket.DataHandler <- err
-					g.Websocket.SetCanUseAuthenticatedEndpoints(false)
+					select {
+					case g.Websocket.DataHandler <- err:
+					default:
+						log.Errorf(log.WebsocketMgr, "%s websocket handle data error: %v", g.Name, err)
+					}
 				}
+			default:
+			}
+			return
+		case resp := <-responseStream:
+			err := g.wsHandleData(resp.Raw)
+			if err != nil {
+				g.Websocket.DataHandler <- err
 			}
 		}
 	}
+}
 
-	return nil
+// wsFunnelConnectionData receives data from multiple connection and pass the data
+// to wsRead through a channel responseStream
+func (g *Gateio) wsFunnelConnectionData(ws stream.Connection) {
+	defer g.Websocket.Wg.Done()
+	for {
+		resp := ws.ReadMessage()
+		if resp.Raw == nil {
+			return
+		}
+		responseStream <- stream.Response{Raw: resp.Raw}
+	}
+}
+
+func (g *Gateio) wsHandleData(respRaw []byte) error {
```

Comment:
I'm not seeing the orderbook processing matching the documentation:
https://www.gate.io/docs/developers/apiv4/ws/en/#order-book-channel
Its important that we have trust in our orderbooks

This style of orderbook processing appears to be the same as the other kinds that you have implemented recently and you can adapt that existing code to suit.

Reply:
I don't think the order book implementation has to be changed as of Kucoin since we can just simply load the snapshot for the first time and continue updating on the consecutive push data messages.
If you think the way I loaded the order book snapshot and updated the change is not accurate ... the implementation is prone to change

Reply:
I'll have a look............... But GateIO is telling us how to maintain their order book. Their [changelog](https://www.gate.io/docs/developers/apiv4/ws/en/#spot-websocket-v4) has not changed since its introduction on 2021-03-17 and I would prefer that we actually follow their instructions